### PR TITLE
feat(prototypes): Allow attaching notification to a target ref in CopyToClipboard prototype

### PR DESCRIPTION
**Pull request by jurokapsiar**
Notification button can be inside of a submenu - after clicking on it, the submenu closes and developer might want to attach the notification to the parent menu item.

### CODE REVIEW

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 1]
```diff
@@ -28,9 +35,21 @@ export const NotificationProvider: React.FC = props => {
 
   return (
     <>
-      <Portal open={!!notification}>
-        <Notification>{notification}</Notification>
-      </Portal>
+      {!!targetRef ? (
+        !!notification && (
+          <Tooltip
+            content={notification}
```
Comment (lucivpav) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]:
```suggestion
            content={{content: notification, variables: {primary: true}}}
```

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 2]
```diff
@@ -28,9 +35,21 @@ export const NotificationProvider: React.FC = props => {
...
+            pointing={true}
+            open={!!notification}
+            target={targetRef.current}
+            variables={{ primary: true }}
```
Comment (lucivpav) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]:
```suggestion
```

Comment (lucivpav) [reply docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: The `primary` prop should be set on `content`, otherwise the appropriate style is not applied:
image

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 3]
```diff
@@ -28,9 +35,21 @@ export const NotificationProvider: React.FC = props => {
 
   return (
     <>
-      <Portal open={!!notification}>
-        <Notification>{notification}</Notification>
-      </Portal>
+      {!!targetRef ? (
+        !!notification && (
+          <Tooltip
```
Comment (lucivpav) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: The following props are missing, yet they are used in `CopyToClipboard.tsx`'s `Tooltip`: align, pointing, position.

File: docs/src/prototypes/CopyToClipboard/index.tsx [hunk 4]
```diff
@@ -27,27 +27,37 @@ const CopyToClipboardPrototype: React.FC<CopyToClipboardPrototypeProps> = props
 }
 
 const CopyToClipboardInMenu: React.FC = props => {
```
Comment (lucivpav) [inline docs/src/prototypes/CopyToClipboard/index.tsx:30]: Let's keep the original example, and add a new example demonstrating the `targetRef` prop.

File: packages/react/src/components/Tooltip/Tooltip.tsx [hunk 5]
```diff
@@ -161,9 +161,9 @@ export default class Tooltip extends AutoControlledComponent<TooltipProps, Toolt
     const { mountNode, children, trigger } = this.props
     const tooltipContent = this.renderTooltipContent(classes.content, rtl, accessibility)
 
-    const triggerElement = React.Children.only(
-      childrenExist(children) ? children : trigger,
-    ) as React.ReactElement<any>
+    const triggerElement =
+      (children || trigger) &&
+      (React.Children.only(childrenExist(children) ? children : trigger) as React.ReactElement<any>)
```
Comment (layershifter) [inline packages/react/src/components/Tooltip/Tooltip.tsx]: It's hard to follow what happens there, can we simplify?

```tsx
    const triggerNode = childrenExist(children) ? children : trigger
    const triggerElement = triggerNode && React.Children.only(triggerNode) as React.ReactElement
    const triggerProps = this.getTriggerProps(triggerElement)
```

File: docs/src/prototypes/CopyToClipboard/index.tsx [hunk 6]
```diff
@@ -27,27 +27,37 @@ const CopyToClipboardPrototype: React.FC<CopyToClipboardPrototypeProps> = props
...
-      ],
-    },
+    render =>
+      render(item, (Component, props) => (
+        <Ref innerRef={targetRef}>
+          <Component {...props} />
+        </Ref>
+      )),
   ]
   return <Menu items={items} />
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/index.tsx:77]: Can keep existing example and add one more instead?

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 7]
```diff
@@ -28,9 +35,21 @@ export const NotificationProvider: React.FC = props => {
...
+            pointing={true}
+            open={!!notification}
+            target={targetRef.current}
+            variables={{ primary: true }}
+          />
+        )
+      ) : (
+        <Portal open={!!notification}>
+          (<Notification>{notification}</Notification>)
+        </Portal>
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: Condition there is very unclear.

Why we show `Notification` or `Tooltip` based on `targetRef`? Should we introduce an additional `position` or prop?

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 8]
```diff
@@ -28,9 +35,21 @@ export const NotificationProvider: React.FC = props => {
...
   return (
     <>
-      <Portal open={!!notification}>
-        <Notification>{notification}</Notification>
-      </Portal>
+      {!!targetRef ? (
+        !!notification && (
+          <Tooltip
+            content={notification}
+            pointing={true}
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: It doesn't respect `pointing` value from the parent component

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 9]
```diff
@@ -28,9 +35,21 @@ export const NotificationProvider: React.FC = props => {
...
+            open={!!notification}
+            target={targetRef.current}
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: This is dangerous, let's use `target`:

```suggestion
            target={targetRef}
```

File: docs/src/prototypes/CopyToClipboard/index.tsx [hunk 10]
```diff
@@ -27,27 +27,37 @@ const CopyToClipboardPrototype: React.FC<CopyToClipboardPrototypeProps> = props
 }
 
 const CopyToClipboardInMenu: React.FC = props => {
+  const targetRef = React.useRef<HTMLElement>(null)
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/index.tsx]: https://github.com/stardust-ui/react/pull/1900/files#r322141672

```suggestion
  const [target, setTarget] = React.useState<HTMLElement>(null)
```

File: docs/src/prototypes/CopyToClipboard/index.tsx [hunk 11]
```diff
@@ -27,27 +27,37 @@ const CopyToClipboardPrototype: React.FC<CopyToClipboardPrototypeProps> = props
...
-                value={'Julius Caesar'}
-                trigger={<Component {...props} />}
-              />
-            )
-          }),
-      ],
-    },
+    render =>
+      render(item, (Component, props) => (
+        <Ref innerRef={targetRef}>
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/index.tsx]: https://github.com/stardust-ui/react/pull/1900/files#r322141672

```suggestion
        <Ref innerRef={setTarget}>
```

File: docs/src/prototypes/CopyToClipboard/index.tsx (hunk 4 above)
Comment (jurokapsiar) [reply docs/src/prototypes/CopyToClipboard/index.tsx:30]: why? we already have the example for not attached, no need to duplicate

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx (hunk 8 above)
Comment (jurokapsiar) [reply docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: according to design, this should never be pointing, so I will always set it to false

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx (hunk 9 above)
Comment (jurokapsiar) [reply docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: wouldn't that be inconsistent with Popup component where we have
```
target?: HTMLElement
```

Comment (layershifter) [reply docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: Oops, wrong suggestion. It should be the same as in `Popup` and be:

```tsx
target={target}
```

So we will pass `target` as `HTMLElement` instead of refs.

***

> Keep in mind that useRef doesn’t notify you when its content changes. Mutating the .current property doesn’t cause a re-render. If you want to run some code when React attaches or detaches a ref to a DOM node, you may want to use a callback ref instead.

https://reactjs.org/docs/hooks-reference.html#useref

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx (hunk 2 above)
Comment (jurokapsiar) [reply docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: the notification should be black according to the spec

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx (hunk 7 above)
Comment (jurokapsiar) [reply docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: changed the API - now there is a separate prop for tooltip (shorthand) and a separate for ataching the notification (TooltipProps or boolean, but we can change it to just subset of tooltip props)

File: docs/src/prototypes/CopyToClipboard/index.tsx [hunk 12]
```diff
@@ -27,27 +26,51 @@ const CopyToClipboardPrototype: React.FC<CopyToClipboardPrototypeProps> = props
...
+    content: 'Edit',
+    menu: [
+      'Open File...',
+      'Save File...',
+      render =>
+        render('Copy text', (Component, props) => {
+          return (
+            <CopyToClipboard
+              attach={{ target: targetRef.current }}
+              value={'Julius Caesar'}
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/index.tsx]:
```suggestion
              value='Julius Caesar'
```

_nit_

File: docs/src/prototypes/CopyToClipboard/index.tsx (hunk 10 above)
Comment (layershifter) [reply docs/src/prototypes/CopyToClipboard/index.tsx]: Oops, was not really applied :(

File: docs/src/prototypes/CopyToClipboard/index.tsx [hunk 13]
```diff
@@ -27,27 +26,51 @@ const CopyToClipboardPrototype: React.FC<CopyToClipboardPrototypeProps> = props
...
+    key: 'edit',
+    content: 'Edit',
+    menu: [
+      'Open File...',
+      'Save File...',
+      render =>
+        render('Copy text', (Component, props) => {
+          return (
+            <CopyToClipboard
+              attach={{ target: targetRef.current }}
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/index.tsx]:
```suggestion
              attach={{ target }}
```

https://github.com/stardust-ui/react/pull/1900/files#r322141672

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 14]
```diff
@@ -1,24 +1,34 @@
...
-  children?: React.ReactNode
+  content: React.ReactNode
+  target?: HTMLElement
+  trigger?: JSX.Element
 }
 
-type NotificationContextValue = (value: React.ReactNode, timeout: number) => void
+type NotificationContextValue = (
+  content: React.ReactNode,
+  targe: HTMLElement | null,
```
Comment (pompomon) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]: Shouldn't it be "target" instead of "targe"?

File: docs/src/prototypes/CopyToClipboard/CopyToClipboard.tsx [hunk 15]
```diff
@@ -1,41 +1,25 @@
...
 import * as React from 'react'
 
-import { NotificationContext } from './NotificationProvider'
+import { Notification, NotificationContext } from './NotificationProvider'
 
 export type CopyToClipboardProps = {
-  attached?: boolean
-  pointing?: boolean
+  tooltip?: ShorthandValue<TooltipProps>
+  attach?: boolean
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/CopyToClipboard.tsx]:
```suggestion
  attached?: boolean
```

Should it be `attached` again?

File: docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx [hunk 16]
```diff
@@ -28,23 +38,38 @@ export const NotificationProvider: React.FC = props => {
...
+
+    if (target || trigger) {
+      return Tooltip.create({ ...tooltipProps, offset: '0 10' })
+    }
+
     return (
-      <div className={classes.root}>
-        <div className={classes.overlay}>
-          <div className={classes.content}>{children}</div>
+      <Portal open={true}>
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/NotificationProvider.tsx]:
```suggestion
      <Portal open>
```

_nit_

File: docs/src/prototypes/CopyToClipboard/themeOverrides.ts [hunk 17]
```diff
@@ -78,44 +52,16 @@ const themeOverrides: ThemeOverrides = {
...
-
-        ...(v.primary && {
-          backgroundImage: pointerSvgUrl(v.primaryContentBackgroundColor),
-        }),
-      }),
-    },
   },
   componentVariables: {
     Notification: (siteVariables): NotificationVariables => ({
       contentBackgroundColor: siteVariables.colorScheme.default.foreground,
```
Comment (layershifter) [inline docs/src/prototypes/CopyToClipboard/themeOverrides.ts]: All these variables can be removed as they are not used.

**Review by layershifter — APPROVED** [review]
